### PR TITLE
Fix TFWav2Vec2 SpecAugment

### PR DESCRIPTION
**Pull request by will-rice**
# What does this PR do?
This PR fixes the SpecAugment implementation for TFWav2Vec2. I had a lot of trouble with this during the original PR, so I'm not 100% if this is correct. I would love feedback because at this point there must be a knowledge gap. 

Fixes # (issue)
https://github.com/huggingface/transformers/issues/12264

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten

### CODE REVIEW

File: src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py [hunk 1]
```diff
@@ -1139,41 +1139,39 @@ def _conv_out_length(input_length, kernel_size, stride):
...
             mask_time_indices = _compute_mask_indices(
                 (batch_size, sequence_length),
-                self.config.mask_time_prob,
-                self.config.mask_time_length,
+                mask_prob=self.config.mask_time_prob,
+                mask_length=self.config.mask_time_length,
                 min_masks=2,
             )
-            hidden_states = tf.tensor_scatter_nd_update(hidden_states, mask_time_indices, self.masked_spec_embed)
+            hidden_states = hidden_states * tf.expand_dims(mask_time_indices, -1)
```
Comment (will-rice) [inline src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py:1165]: This is wrong. The indices represent with 1's the position to mask, so this essentially zero's out everything not in the mask window. If I still wanted to multiply it I would need to invert the mask.

Comment (will-rice) [reply src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py:1165]: it should be this [op](https://www.tensorflow.org/api_docs/python/tf/tensor_scatter_nd_update), but I'm not doing it correctly. The indices tests pass, so I'm probably not passing the indices in the right format.

**Comment by will-rice** [comment]
Based on the original paper the time mask should be applied on the time axis like 
image
However, wav2vec2 is masking hidden states. In principle, it should work the same way and if correctly applied you should be able to see the zeroed span in the time dimension of the hidden states.

File: src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py [hunk 2]
```diff
@@ -1220,9 +1221,11 @@ def call(
 
         mask_time_indices = kwargs.get("mask_time_indices", None)
         if mask_time_indices is not None:  # apply SpecAugment along time axis with given indices
-            hidden_states = tf.tensor_scatter_nd_update(hidden_states, mask_time_indices, self.mask_spec_embed)
+            mask = tf.cast(tf.logical_not(tf.cast(mask_time_indices, tf.bool)), tf.float32)
+            hidden_states = hidden_states * tf.expand_dims(mask, -1)
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py:1225]: I'm a bit confused here - as it's implemented now this would just zero out the masked hidden states no? But we should set the vectors to `self.mask_spec_embed` instead. Think it makes sense to use tf.where here and cast `self.mask_spec_embed` to be of the same size as `hidden_states`, so:

```
hidden_states = tf.where(mask_time_indices, expanded_mask_spec_embed, hidden_states)
```

**Comment by patrickvonplaten** [comment]
Hey @will-rice,

Thanks for fixing the bug. As I understand it, when applying spec_augment along the time axis we are actually not setting the values to zero but to a trained mask embedding vector (this is mostly because that's how it's used for pretraining actually). 
IMO, to correctly implement this one has to make use of `tf.where`. A good way would be to expand the masked indices and `self.masked_spec_embed` to be of the same size as `hidden_states` (as explained above) ati use `tf.where`.

For spec_augment along the feature axis I would also suggest to use `tf.where` -> expand the feature indices (this time along the time axis (seq length) and then one can simply do:

```
hidden_states = tf.where(expanded_mask, hidden_states, 0)
```

**Comment by patrickvonplaten** [comment]
Let me know if this is not understandable ;-)

**Comment by will-rice** [comment]
> Hey @will-rice,
> 
> Thanks for fixing the bug. As I understand it, when applying spec_augment along the time axis we are actually not setting the values to zero but to a trained mask embedding vector (this is mostly because that's how it's used for pretraining actually).
> IMO, to correctly implement this one has to make use of `tf.where`. A good way would be to expand the masked indices and `self.masked_spec_embed` to be of the same size as `hidden_states` (as explained above) ati use `tf.where`.
> 
> For spec_augment along the feature axis I would also suggest to use `tf.where` -> expand the feature indices (this time along the time axis (seq length) and then one can simply do:
> 
> ```
> hidden_states = tf.where(expanded_mask, hidden_states, 0)
> ```

"fixing" :stuck_out_tongue:. I believe I understand the way to do it now but will post questions here if I get stuck. Thank you for walking through this!

**Review by patrickvonplaten — APPROVED** [review]
This looks very nice - thanks for fixing it!
